### ISSUE TITLE
De-couple source and destination from filesystem

### ISSUE THREAD

**Issue by bignose-debian** [brechtm/rinohtype#176]
The primary Python API for ricohtype (in particular, `rinoh.frontend.rst` and `rinoh.document.Document` assumes reading from files and rendering to files, with paths on the filesystem.

This is not a necessary part of converting formats from one to another, and it makes it needlessly difficult to make rinoh part of a larger flow within an application.

I propose changing the API so that the API does *not* primarily assume every parsing is done from a file, and does *not* primarily assume every render will be to an output file.

Docutils, for example, parses from a text (and it is up to the caller how to get that text), and does not assume any files.

**Comment by bignose-debian** [comment]
I would expect, for example to be able to use rinohtype this way in a Python application:

```
import rinoh.frontend.rst
import rinoh.templates

# Generate the reST content somehow.
source_document = """Lorem ipsum, **dolor** sit amet."""

parser = rinoh.frontend.rst.ReStructuredTextReader()
document_tree = parser.parse(source_document)

template = rinoh.template.Article()
pdf_content = template.render(document_tree)

# ... continue processing the PDF content ...
```

**Comment by brechtm** [comment]
You are absolutely correct, of course. I always planned to also offer a friendly API. Due to lack of time, the focus is still on the command line renderer _rinoh_ and the Sphinx builder.

A pull request would be highly appreciated! I don't have a strong idea about what the API should look like in this case, so feel free to suggest something.
